### PR TITLE
feat: async update artifact pull time and repository pull count

### PR DESCRIPTION
**Pull request by chlins**
Fixes: https://github.com/goharbor/harbor/issues/15963

Signed-off-by: chlins <chenyuzh@vmware.com>

### CODE REVIEW

File: src/controller/event/handler/internal/artifact.go [hunk 1]
```diff
@@ -16,19 +16,63 @@ package internal
...
 	"github.com/goharbor/harbor/src/controller/tag"
+	"github.com/goharbor/harbor/src/lib/config"
 	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/lib/orm"
 	"github.com/goharbor/harbor/src/lib/q"
 )
 
+const (
+	// defaultAsyncFlushDuration is the default flush interval.
+	defaultAsyncFlushDuration = 10 * time.Second
```
Comment (wy65701436) [inline src/controller/event/handler/internal/artifact.go:38]: can it be configurated?

File: src/controller/event/handler/internal/artifact.go [hunk 2]
```diff
@@ -55,43 +99,137 @@ func (a *Handler) IsStateful() bool {
...
 
 func (a *Handler) onPull(ctx context.Context, event *event.ArtifactEvent) error {
+	// if duration is equal to 0 or negative, keep original sync mode.
+	if asyncFlushDuration <= 0 {
+		var tagName string
+		if len(event.Tags) > 0 {
+			tagName = event.Tags[0]
+		}
+
+		if !config.PullTimeUpdateDisable(ctx) {
```
Comment (wy65701436) [inline src/controller/event/handler/internal/artifact.go:109]: do we still have to keep PullTimeUpdateDisable & PullCountUpdateDisable?

cc @stonezdj

File: src/controller/event/handler/internal/artifact.go [hunk 3]
```diff
@@ -55,43 +99,137 @@ func (a *Handler) IsStateful() bool {
...
+	defer a.pullCountLock.Unlock()
+
+	a.pullCountStore[event.Artifact.RepositoryID] = a.pullCountStore[event.Artifact.RepositoryID] + 1
+}
+
+func (a *Handler) syncFlushPullTime(ctx context.Context, artifactID int64, tagName string, time time.Time) {
+	var tagID int64
+
+	if tagName != "" {
 		tags, err := tag.Ctl.List(ctx, &q.Query{
```
Comment (wy65701436) [inline src/controller/event/handler/internal/artifact.go:169]: q.New()?

File: src/controller/event/handler/internal/artifact.go [hunk 4]
```diff
@@ -55,43 +99,137 @@ func (a *Handler) IsStateful() bool {
...
 		} else {
 			if len(tags) != 0 {
 				tagID = tags[0].ID
 			}
 		}
 	}
-	if err := artifact.Ctl.UpdatePullTime(ctx, event.Artifact.ID, tagID, time.Now()); err != nil {
-		log.Debugf("failed to update pull time form artifact %d, %v", event.Artifact.ID, err)
+
+	if err := artifact.Ctl.UpdatePullTime(ctx, artifactID, tagID, time); err != nil {
```
Comment (wy65701436) [inline src/controller/event/handler/internal/artifact.go:184]: It seems that artifact.Ctl.UpdatePullTime() doesn't handle well the situation of tagID equals 0, please update it to handle the ErrorNotFound.

File: src/controller/event/handler/internal/artifact.go [hunk 5]
```diff
@@ -55,43 +99,137 @@ func (a *Handler) IsStateful() bool {
...
+		log.Warningf("failed to update pull time form artifact %d, %v", artifactID, err)
```
Comment (wy65701436) [inline src/controller/event/handler/internal/artifact.go:185]: form ==> for?

File: src/controller/event/handler/internal/artifact.go [hunk 6]
```diff
@@ -55,43 +99,137 @@ func (a *Handler) IsStateful() bool {
...
-	if err := repository.Ctl.AddPullCount(ctx, event.Artifact.RepositoryID); err != nil {
-		log.Debugf("failed to add pull count repository %d, %v", event.Artifact.RepositoryID, err)
+func (a *Handler) asyncFlushPullTime(ctx context.Context) {
+	for {
+		<-time.After(asyncFlushDuration)
+		a.pullTimeLock.Lock()
+
+		for key, time := range a.pullTimeStore {
+			keys := strings.Split(key, ":")
+			artifactID, err := strconv.ParseInt(keys[0], 10, 64)
```
Comment (wy65701436) [inline src/controller/event/handler/internal/artifact.go:202]: check keys length before using?

File: src/controller/event/handler/internal/artifact.go [hunk 7]
```diff
@@ -55,43 +99,137 @@ func (a *Handler) IsStateful() bool {
...
 
-func (a *Handler) addPullCount(ctx context.Context, event *event.ArtifactEvent) {
-	if err := repository.Ctl.AddPullCount(ctx, event.Artifact.RepositoryID); err != nil {
-		log.Debugf("failed to add pull count repository %d, %v", event.Artifact.RepositoryID, err)
+func (a *Handler) asyncFlushPullTime(ctx context.Context) {
+	for {
+		<-time.After(asyncFlushDuration)
+		a.pullTimeLock.Lock()
+
+		for key, time := range a.pullTimeStore {
```
Comment (wy65701436) [inline src/controller/event/handler/internal/artifact.go:200]: can we consider batch update?

File: src/controller/event/handler/internal/artifact.go [hunk 8]
```diff
@@ -16,19 +16,63 @@ package internal
...
+)
+
+var (
+	asyncFlushDuration time.Duration
+)
+
+func init() {
+	// get async flush duration from env, if not provide,
+	// use default value: 10*time.second
+	envDuration := os.Getenv("ARTIFACT_PULL_TIME_ASYNC_FLUSH_DURATION")
```
Comment (wy65701436) [inline src/controller/event/handler/internal/artifact.go:48]: change another name as it will handle pull count as well?

**Comment by wy65701436** [comment]
@chlins please add the situation description that we discussed offline.

File: src/controller/event/handler/internal/artifact.go (hunk 2 above)
Comment (reasonerjt) [reply src/controller/event/handler/internal/artifact.go:109]: I hope not.

File: src/controller/event/handler/internal/artifact.go [hunk 9]
```diff
@@ -55,43 +99,137 @@ func (a *Handler) IsStateful() bool {
 }
 
 func (a *Handler) onPull(ctx context.Context, event *event.ArtifactEvent) error {
+	// if duration is equal to 0 or negative, keep original sync mode.
```
Comment (reasonerjt) [inline src/controller/event/handler/internal/artifact.go:102]: Does this func always return nil?

File: src/controller/event/handler/internal/artifact.go (hunk 4 above)
Comment (reasonerjt) [reply src/controller/event/handler/internal/artifact.go:184]: seems we should handle that in the artifact controller?

File: src/pkg/artifact/dao/dao.go [hunk 10]
```diff
@@ -180,7 +180,27 @@ func (d *dao) Update(ctx context.Context, artifact *Artifact, props ...string) e
...
+			if artifact.PullTime.Before(old.PullTime) {
+				artifact.PullTime = old.PullTime
+				continue
+			}
+		}
+
+		newProps = append(newProps, prop)
+	}
+
+	n, err := ormer.Update(artifact, newProps...)
```
Comment (heww) [inline src/pkg/artifact/dao/dao.go:203]: I prefer to add a new method to only update the pull time in the DAO.

File: src/controller/event/handler/internal/artifact.go (hunk 8 above)
Comment (chlins) [reply src/controller/event/handler/internal/artifact.go:48]: ARTIFACT_ASYNC_FLUSH_DURATION

File: src/controller/event/handler/internal/artifact.go (hunk 9 above)
Comment (chlins) [reply src/controller/event/handler/internal/artifact.go:102]: > Does this func always return nil?

Yes, it always returns nil and only print error logs if meet error.

File: src/controller/event/handler/internal/artifact.go (hunk 3 above)
Comment (chlins) [reply src/controller/event/handler/internal/artifact.go:169]: ok

File: src/controller/event/handler/internal/artifact.go (hunk 4 above)
Comment (chlins) [reply src/controller/event/handler/internal/artifact.go:184]: Let's use another pr to fix the situation.

File: src/pkg/artifact/dao/dao.go (hunk 10 above)
Comment (chlins) [reply src/pkg/artifact/dao/dao.go:203]: good point

File: src/controller/event/handler/internal/artifact.go (hunk 7 above)
Comment (chlins) [reply src/controller/event/handler/internal/artifact.go:199]: keep for more investigation.

**Review by wy65701436 — APPROVED** [review]
lgtm

File: src/controller/event/handler/internal/artifact.go (hunk 2 above)
Comment (chlins) [reply src/controller/event/handler/internal/artifact.go:109]: use other pr to remove it if we need.

File: src/pkg/artifact/dao/dao.go [hunk 11]
```diff
@@ -189,6 +193,27 @@ func (d *dao) Update(ctx context.Context, artifact *Artifact, props ...string) e
...
+func (d *dao) UpdatePullTime(ctx context.Context, id int64, pullTime time.Time) error {
+	ormer, err := orm.FromContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	// can only be retained to the second if not format
+	formatPullTime := pullTime.Format("2006-01-02 15:04:05.999999")
+	// update db only if pull_time is null or pull_time < (in-coming)pullTime
+	sql := "UPDATE artifact SET pull_time = ? WHERE id = ? AND pull_time IS NULL OR pull_time < ?"
```
Comment (heww) [inline src/pkg/artifact/dao/dao.go:206]: `"UPDATE artifact SET pull_time = ? WHERE id = ? AND pull_time IS NULL OR pull_time < ?"` => 

`"UPDATE artifact SET pull_time = ? WHERE id = ? AND (pull_time IS NULL OR pull_time < ?)"`

File: src/controller/event/handler/internal/artifact.go [hunk 12]
```diff
@@ -16,19 +16,63 @@ package internal
...
+	} else {
+		duration, err := strconv.ParseInt(envDuration, 10, 64)
+		if err != nil {
+			log.Warningf("error to parse ARTIFACT_PULL_ASYNC_FLUSH_DURATION: %v, will use default value: %v", err, defaultAsyncFlushDuration)
+			asyncFlushDuration = defaultAsyncFlushDuration
+		} else {
+			asyncFlushDuration = time.Duration(duration) * time.Second
+		}
+	}
+}
```
Comment (heww) [inline src/controller/event/handler/internal/artifact.go:61]: Should this value be configurable in the system configuration? @wy65701436

File: src/controller/event/handler/internal/artifact.go [hunk 13]
```diff
@@ -55,43 +99,136 @@ func (a *Handler) IsStateful() bool {
...
+	if err := repository.Ctl.AddPullCount(ctx, repositoryID, count); err != nil {
+		log.Warningf("failed to add pull count repository %d, %v", repositoryID, err)
 	}
-	return
 }
 
-func (a *Handler) addPullCount(ctx context.Context, event *event.ArtifactEvent) {
-	if err := repository.Ctl.AddPullCount(ctx, event.Artifact.RepositoryID); err != nil {
-		log.Debugf("failed to add pull count repository %d, %v", event.Artifact.RepositoryID, err)
+func (a *Handler) asyncFlushPullTime(ctx context.Context) {
```
Comment (wy65701436) [inline src/controller/event/handler/internal/artifact.go:194]: generally, we should consider the listen os shutdown signal in the goroutine to ensure update?

File: src/controller/event/handler/internal/artifact.go (hunk 12 above)
Comment (chlins) [reply src/controller/event/handler/internal/artifact.go:61]: Use another pr to expose this to system configuration if we need.

File: src/controller/event/handler/internal/artifact.go (hunk 2 above)
Comment (stonezdj) [reply src/controller/event/handler/internal/artifact.go:109]: We should keep the setting of PullAuditLogDisable

**Review by stonezdj — APPROVED** [review]
LGTM

**Pull request merged**